torch/autograd/graph.py

```diff
@@ -269,3 +271,125 @@ def __setstate__(self, state):
+            else:
+                # Store an additional handle
+                handle = _tid_to_weakhandle[tid]()
+            return _ctx_id, handle
+
+        def unpack_hook(tup):
+            ctx_id, handle = tup
+            assert ctx_id == _ctx_id, (
+                "Trying to backward outside of the 'allow_mutation_on_saved_tensors' context"
+                "in which the graph was originally recorded.")
+            if handle in _cloned:
+                res = _cloned[handle]
+            else:
+                res = _original[handle]
+            return res
+
+        super().__init__(pack_hook, unpack_hook)
+
+class _CloneArgBeforeMutateMode(TorchDispatchMode):
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+
+        # Bug? Maybe something to do with wrapped number + dispatcher, not sure why yet
+        # We don't really need this unless we try to print though, but keeping for now
+        if type(args[0]) is int:
+            args = [torch.tensor(args[0]), *args[1:]]
+
+        # (only for in-place ops now, we may want to handle out= later)
+        if func.__name__.split('.')[0][-1] == "_":
+            # The first argument is assumed to be modified in-place
+            tid = _get_tid(args[0])
+            sid = _get_sid(args[0])
+            if sid in _sid_to_tid:
+                for tid in _sid_to_tid[sid]:
+                    if tid not in _tid_to_weakhandle:
+                        # It's never been saved
+                        continue
+                    handle = _tid_to_weakhandle[tid]()
+                    if handle is None or handle in _cloned:
+                        # It's been saved, but backward was run OR
+                        # The same exactly tensor has been cloned already
+                        continue
+                    _cloned[handle] = _original[handle].clone()
+                    del _original[handle]
+            else:
+                # this can happen with math views, I'm not sure why yet
+                assert not args[0]._is_view()
+
+        rs = func(*args, **kwargs)
+        return rs
+
+@contextlib.contextmanager
+def allow_mutation_on_saved_tensors():
+    """Context manager under which mutating tensors saved for backward is allowed
+
+    Under this context manager, tensors saved for backward are cloned on mutation,
+    so the original version can still be used during backward. Normally, mutating a tensor
+    saved for backward will result in an error raised when it's used during backward.
+    """
+    global _inside_ctx, _ctx_id
+    with _swap_with_cloned(), _CloneArgBeforeMutateMode():
+        try:
+            _ctx_id += 1
+            if _inside_ctx:
+                raise RuntimeError("allow_mutation_on_saved_tensors contexts cannot be nested")
+            _inside_ctx = True
+            yield
+        finally:
+            _cloned.clear()
+            _ctx_id += 1
```

Comment:
-= ?

Reply:
Not actually a typo haha, I do want to increment here so that I can detect when I reenter into a new context, the _ctx_id would be different.